### ISSUE TITLE
Southern Africa gets matched to South Africa and shouldn't

### ISSUE THREAD

**Issue by kendonB** [vincentarelbundock/countrycode#71]
All in the title

**Comment by kendonB** [comment]
Also: 
- "Sub-Saharan Africa" gets matched to "Western Sahara"
- "Micronesia" gets matched to "Micronesia, Federated States of"
- "Australia/New Zealand" gets matched to "New Zealand"

**Comment by kendonB** [comment]
These are all from https://esa.un.org/unpd/wpp/DVD/Files/1_Indicators%20(Standard)/ASCII_FILES/WPP2015_DB02_Populations_Annual.zip

**Comment by vincentarelbundock** [comment]
Thanks for the report!

The two related problems should be fixed.

I haven't seen Micronesia data much. Do international organizations usually refer to the federated States?

I don't think I want to fix the australis one. The goal of country code cannot be to correctly translate every non-standard input, otherwise I could have to craft reflexes.to guard against every Canada/USA compounds out there. That's what the warning flag is there for.

Makes sense?

**Comment by vincentarelbundock** [comment]
"africa-related"

**Comment by vincentarelbundock** [comment]
Craft regex to guard

**Comment by kendonB** [comment]
Micronesia is the region: https://en.m.wikipedia.org/wiki/Micronesia
Federated States is a country: https://en.m.wikipedia.org/wiki/Federated_States_of_Micronesia

I don't think I've ever seen "Micronesia" refer to the federated states, but it's possible probably.

I think there probably is an elegant solution that solves the Australia/New Zealand problem, but I don't have it...

I haven't seen any warnings. I'll update to the development version.

**Comment by vincentarelbundock** [comment]
you need to use the ```warn=TRUE``` flag to see warnings. Perhaps the default should be TRUE, but that can lead to LOTS of warnings when converting a long vector.

What about the argument that COUNTRYcode expects countries (as opposed to regions) as inputs? My reflex would be to say this isn't such an unreasonable constraint for a package which aims to do a simple task...

**Comment by kendonB** [comment]
The default should probably be TRUE. And you can get countrycode to only act on unique(vector) to prevent lots of warnings, and you'll get a big speed bump.

I agree I don't think it's too unreasonable to expect users to filter out regions. But, many UN and World Bank datasets still come with aggregates/non-countries that aren't flagged as such.

Perhaps a flag_aggregate function, that returns a ``lgl`` when detecting the UN and WDI aggregate names, would be a good addition to the package.

**Comment by vincentarelbundock** [comment]
https://github.com/vincentarelbundock/countrycode/issues/72

Aggregate detection may be a useful feature, but where do we stop? What other strings should we exclude? Regions, continents, "Fiji" branded water?

In principle, I like the idea. The main constraint is that I have basically zero free time to invest, and that I don't really use the package anymore (so no real itch to scratch on my end, beyond duty to maintain).

**Comment by kendonB** [comment]
I'm open an issue - someone will implement it (probably me). I think handling common datasets from the big organizations is a start.

**Comment by vincentarelbundock** [comment]
In general, I think we should limit the number of arguments to the main function; we don't want to complicate the API with thousands of switches. Would it be possible to just add an "aggregate" column to the CSV file with appropriate regexes?

**Comment by kendonB** [comment]
I agree. I imagined it would be a separate function and work something like this:

```
.data %>% 
  # detect_aggr returns a lgl
  filter(!detect_aggr(country)) %>% 
  # Now should be mostly free of aggregates and this should work
  mutate(iso3c = countrycode(country, "country.name", "iso3c"))
```
As far as I know, none of the coding systems have codes for aggregates? So it seems unnecessary to include them in the main table.

**Comment by vincentarelbundock** [comment]
I like this. Maybe we should expand the idea to cover the other regional corner cases that you pointed out earlier (e.g., "southern africa").

**Comment by cjyetman** [comment]
@kendonB Why not use countrycode to convert all of the country and region names and then filter out those that are `NA`, i.e. were not able to be converted by countrycode? Then the strategy for countrycode would be to make sure that it doesn't incorrectly match common region/aggregate names as country names.

That is what I have often done in the past working with WDI data, for example:
```
.data %>% 
  mutate(iso3c = countrycode(country, "country.name", "iso3c")) %>% 
  filter(!is.na(iso3c))
```

**Comment by cjyetman** [comment]
Another potential solution would be a good use case for the new custom dictionary feature in the dev version of countrycode. (and with the just merged PR #77, the below example will run a lot faster)

I found [here](https://esa.un.org/unpd/wpp/Download/Other/Documentation/) a link to an XLS file that contains a "Comprehensive list of locations with codes..." on the UN WPP website. That [XLS file](https://esa.un.org/unpd/wpp/DVD/Files/4_Other%20Files/WPP2015_F01_LOCATIONS.XLS) can be used to build a custom dictionary for countrycode that can do what you want.

For example...
```
library(dplyr)
library(readxl)
library(readr)
library(countrycode)

xlsurl <- 'https://esa.un.org/unpd/wpp/DVD/Files/4_Other%20Files/WPP2015_F01_LOCATIONS.XLS'  
tempxls <- file.path(tempdir(), basename(xlsurl))
download.file(xlsurl, destfile = tempxls)
cntrylkp <- readxl::read_excel(tempxls, skip = 16)
names(cntrylkp)[2] <- 'regionname'  # make easy-to-use column names
names(cntrylkp)[7] <- 'regiontype'  # make easy-to-use column names
cntrylkp <- as.data.frame(cntrylkp)  # make sure it's a data.frame, not a tibble
unlink(tempxls)

# same example file given above
csvurl <- 'https://esa.un.org/unpd/wpp/DVD/Files/1_Indicators%20(Standard)/ASCII_FILES/WPP2015_DB02_Populations_Annual.zip'
tempcsv <- file.path(tempdir(), basename(csvurl))
download.file(csvurl, destfile = tempcsv)
df <- readr::read_csv(tempcsv)
unlink(tempcsv)

df %>% 
  mutate(regiontype = countrycode(Location, 'regionname', 'regiontype', dictionary = cntrylkp)) %>% 
  filter(regiontype == 'Country/Area') %>% 
  mutate(iso3c = countrycode(Location, 'country.name', 'iso3c'))
```

**Comment by vincentarelbundock** [comment]
cdb6b0200adf47
